conftest.py

```diff
@@ -1,35 +1,296 @@
+    - Class of object (DataFrame, Series, Index)
+    - Dtype
+    - Nullability
+    - Size (rows for all, rows/columns for DataFrame)
 
-from utils import make_col, make_frame
+One such fixture is a series of nullable integer data. Given that we generally
+want data across different sizes, we parametrize all fixtures across different
+numbers of rows rather than generating separate fixtures for each different
+possible number of rows. The number of columns is only relevant for DataFrame.
 
+While this core set of fixtures means that any benchmark can be run for any
+combination of these parameters, it also means that we would effectively have
+to parametrize our benchmarks with many fixtures. Not only is parametrizing
+tests with fixtures in this manner unsupported by pytest, it is also an
+inelegant solution leading to cumbersome parameter lists that must be
+maintained across all tests. Instead we make use of the
+`pytest_cases <https://smarie.github.io/python-pytest-cases/>_` pytest plugin,
+which supports the creation of fixture unions: fixtures that result from
+combining other fixtures together. The result is a set of well-defined fixtures
+that allow us to write benchmarks that naturally express the set of objects for
+which they are valid, e.g. `def bench_sort_values(frame_or_index)`.
 
-@pytest.fixture(params=product([100, 10000], [True, False]))
-def col(request):
-    """Create a cudf column.
+The generated fixtures are named according to the following convention:
+`classname_dtype_{dtype}[_nulls_{true|false}][_cols_{num_cols}]`
+where classname is one of the following: index, series, dataframe, indexedframe,
+frame, frame_or_index. Note that in the case of
+indexes, to match Series/DataFrame we simply set `classname=index` and rely on
+the `dtype_{dtype}` component to delineate which index class is actually in use.
 
-    The two parameters are `nrows` and `has_nulls`
-    """
-    return make_col(*request.param)
+In addition to the above fixtures, we also provide the following more specialized
+fixtures:
+    - rangeindex: Since RangeIndex always holds int64 data we cannot conflate
+      it with index_dtype_int64 (a true Int64Index), and it cannot hold nulls.
+      As a result, it is provided as a separate fixture.
+"""
 
 
-@pytest.fixture(params=[100, 10000])
-def df(request):
-    """Create a cudf DataFrame.
+import os
+import re
+import string
+import sys
+from collections.abc import MutableSet
+from functools import partial
+from itertools import groupby
 
-    The two parameters are `nrows`
-    """
-    return make_frame(ncols=5, nkey_cols=0, nrows=request.param)
+import pytest_cases
+
+# TODO: Rather than doing this path hacking (including the sessionstart and
+# sessionfinish hooks), we could just make the benchmarks a (sub)package to
+# enable relative imports. A minor change to consider when these are ported
+# into the main repo.
+sys.path.insert(0, os.path.join(os.getcwd(), "common"))
```

Comment:
Should this use the current working directory, or something relative to `__file__`? I suspect we want the path to be relative to `__file__` so that this suite of benchmarks can be executed from other directories.

(This may be resolved by the TODO / making this a (sub)package.)

Reply:
It would indeed be resolved by the subpackage question, which to me is something that we can resolve when we migrate this repo into cudf proper. If we think that making it a subpackage is the way to go (basically just inserting a `__init__.py` file into the benchmarks/ dir) then I don't think we need to overthink this now. If we don't like the subpackage approach, then we should probably change this path calculation to something more robust like you suggest.

Reply:
I just googled for better solutions and it turns out that [this is also what a lot of other people have come up with](https://stackoverflow.com/questions/33508060/create-and-import-helper-functions-in-tests-without-creating-packages-in-test-di), so I'm not sure that we really _need_ to change this with the subpackage approach. It really is pytest's oversight not providing a nicer way to handle this IMO.